### PR TITLE
Add add and adc instructions

### PR DESCRIPTION
**Pull request by kharland**

### CODE REVIEW

File: arm7_tdmi/lib/src/instruction/arithmetic.dart [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    Arm7TdmiCondition condition,
+    int rd,
+    int rn,
+    int shifterOperand,
+    bool updatesSpsr}) {
+  final op1 = int32.mask(gprs.get(rn));
+  final op2 = int32.mask(shifterOperand + (gprs.cpsr.c ? 1 : 0));
+  final trueResult = op1 + op2;
+
+  if (updatesSpsr && rd == 15) {
```
Comment (matanlurey) [inline arm7_tdmi/lib/src/instruction/arithmetic.dart]: Can we pull out the magic number (15) to something?

Comment (matanlurey) [inline arm7_tdmi/lib/src/instruction/arithmetic.dart]: (Here and below)

File: arm7_tdmi/lib/src/instruction/arithmetic.dart [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+void add(
+    {Arm7TdmiOperatingMode mode,
+    Arm7TdmiRegisters gprs,
+    Arm7TdmiCondition condition,
+    int rd,
+    int rn,
+    int shifterOperand,
+    bool updatesSpsr}) {
+  if (condition.passes(gprs.cpsr)) {
```
Comment (matanlurey) [inline arm7_tdmi/lib/src/instruction/arithmetic.dart]: How about:

```dart
if (!condition.passes(gprs.cpsr)) {
  return;
}
```

File: binary/lib/binary.dart [hunk 3]
```diff
@@ -233,6 +233,34 @@ class Integral implements Comparable<Integral> {
     return result;
   }
 
+  /// Returns 0 if [number] is non-negative, else 1.
+  int sign(int number) {
```
Comment (matanlurey) [inline binary/lib/binary.dart]: What do you think of `bool isSigned(int number)` instead?

File: binary/lib/binary.dart [hunk 4]
```diff
@@ -233,6 +233,34 @@ class Integral implements Comparable<Integral> {
...
+    assert(() {
+      if (!isSigned) {
+        throw new UnsupportedError('Cannot compute sign for unsinged value');
```
Comment (matanlurey) [inline binary/lib/binary.dart]: nit: spelling of _unsigned_

File: binary/lib/binary.dart [hunk 5]
```diff
@@ -233,6 +233,34 @@ class Integral implements Comparable<Integral> {
...
+      }
+      return true;
+    });
+    return getBit(number, length - 1);
+  }
+
+  /// Returns [1] if the [result] of an addition produced a carry bit, else 0.
```
Comment (matanlurey) [inline binary/lib/binary.dart]: WDUT you think of:

```dart
/// Returns `true` if the [result] of an additional produced a carry bit, else `false`.
bool isCarry(int result) => result > max;
```

File: binary/lib/binary.dart [hunk 6]
```diff
@@ -233,6 +233,34 @@ class Integral implements Comparable<Integral> {
...
+  ///
+  /// A carry bit is produced during a two's compliment addition if the unmasked
+  /// result is greater than [Integral.max].
+  int carryFrom(int result) => result > max ? 1 : 0;
+
+  /// Returns 1 if [op1] + [op2] produced a signed overflow in [result], else 0.
+  int overflowFromAdd(int op1, int op2, int result) =>
```
Comment (matanlurey) [inline binary/lib/binary.dart]: Same as above, `bool`?

Also how about `isAddOverflow`?

File: binary/lib/binary.dart [hunk 7]
```diff
@@ -233,6 +233,34 @@ class Integral implements Comparable<Integral> {
...
+      sign(op1) == sign(op2) && sign(result) != sign(op1) ? 1 : 0;
+
+  /// Returns 1 if [op1] - [op2] produced a signed overflow in [result], else 0.
+  int overflowFromSub(int op1, int op2, int result) =>
```
Comment (matanlurey) [inline binary/lib/binary.dart]: `isSubOverflow`? `bool`?

File: binary/lib/binary.dart (hunk 3 above)
Comment (kharland) [inline binary/lib/binary.dart]: Was planning to clean that up in another PR. can change in this one if you'd like.

File: binary/lib/binary.dart (hunk 5 above)
Comment (kharland) [inline binary/lib/binary.dart]: same as above

File: binary/lib/binary.dart (hunk 3 above)
Comment (kharland) [inline binary/lib/binary.dart]: changed the other bools for now. This class already declares a getter isSigned.

File: binary/lib/binary.dart (hunk 5 above)
Comment (kharland) [inline binary/lib/binary.dart]: done

File: binary/lib/binary.dart (hunk 4 above)
Comment (kharland) [inline binary/lib/binary.dart]: done

**Comment by kharland** [comment]
Closed original PR since this one contains both sets of changes

**Comment by kharland** [comment]
Looks like I need to publish binary before this can work since we aren't using the local path: ... dependency anymore. Splitting into two PRs

**Pull request closed without merging**